### ISSUE TITLE
Regression in StreamPipeReader.CopyToAsync() causes part of already consumed content to be copied again

### ISSUE THREAD

**Issue by KrzysFR** [dotnet/runtime#57806]
Using the latest version of `System.IO.Pipelines` (v6.0.0-preview.7.21377.19), the implementation of `StreamPipeReader.CopyToAsync()` always copies from the start of the current segment buffer, even if the pipe was advanced prior to the call. This means that already parsed data is copied again in the output stream, causing issue with parsers.

I think this is a regression introduced by #52159 when it added a new implementation of `CopyToAsync()`

### Description

Here is a reproduction: the code reads from a buffer, skips the first 12 bytes and `CopyToAsync` the rest of pipe into a temporary stream. In v5.0.1 the destination stream does not include the already consumed 12 bytes (expected), but in 6.0.0-preview.7.21377.19, the entire original buffer is copied including the 12 bytes (unexpected).

```c#
// original buffer
var bytes = Encoding.UTF8.GetBytes("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"); // original = 36 bytes
using var source = new MemoryStream(bytes);

// create a pipe reader over this source
var pipe = PipeReader.Create(source);

// simulate parsing a 12-byte header from the stream
var readResult = await pipe.ReadAsync();
// ...do some parsing here, advance by 12 bytes
pipe.AdvanceTo(readResult.Buffer.GetPosition(12), readResult.Buffer.End);

// now copy the rest to another MemoryStream
using var destination = new MemoryStream();
await pipe.CopyToAsync(destination);

// expected: "MNOPQRSTUVWXYZ0123456789" (24 bytes copied, skipping first 12)
Console.WriteLine("Result: " + Encoding.UTF8.GetString(destination.ToArray()));

// OK: System.IO.Pipelines v5.0.1 => "MNOPQRSTUVWXYZ0123456789"
// KO: System.IO.Pipelines v6.0.0-preview.7.21377.19 => "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
```

### Configuration

Tested on .NET SDK 5.0.1 and 6.0-preview7, Windows x64.

Using either .NET 5.0 or 6.0 does not change the result: it is dependent on the version of the NuGet package

### Regression?

Works as intended with System.IO.Pipelines v5.0.1

### Other information

The new implementation of StreamPipeReader.CopyToAsync() does not read the value of `m_readIndex` while copying the first buffered segment. In the repro above, `m_readIndex` is `12`. The code should probably slice the current segment before copying it.

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs#L456-L461

The previous implementation did use the Start position of the buffer to perform the copy.

**Comment by stephentoub** [comment]
cc: @davidfowl, @manandre

**Comment by davidfowl** [comment]
Good catch! Seems like something we should fix...

**Comment by KrzysFR** [comment]
That was fast! 👍 

If I understand the PR correctly, the fix will be included in 6.0-rc1 ?

**Comment by davidfowl** [comment]
Yep.
